### ISSUE TITLE
Expanding the functionality of $this->db->like() and it's related methods without breaking anything.

### ISSUE THREAD

**Issue by Nudasoft** [bcit-ci/CodeIgniter#5641]
Feature request 01
============
Now $this->db->like() method and it's related methods, e.i: this->db->or_like() and etc... support for 3 different wildcard (%) placements. I'm asking for another extra one. lets call it 'none' for now.

The expected result should be when using 'none' wildcard placement option:
$this->db->like('title', 'match', 'none'); 
Should produces: WHERE `title` LIKE 'match' ESCAPE '!'

Reason to request this feature
----------------------------------
MySQL return rows that seemingly do not match the WHERE clause. Please refer this url to get a good idea about what I'm taling here:
[https://stackoverflow.com/questions/15715719/why-does-mysql-return-rows-that-seemingly-do-not-match-the-where-clause](https://stackoverflow.com/questions/15715719/why-does-mysql-return-rows-that-seemingly-do-not-match-the-where-clause)

By using LIKE operator without wildcard character we can get the exact result set that we want.

This is how I do it for now. but dont know if this way of doing this intorduce any security issues. 
$this->db->where('title LIKE ', 'match);

And I think that introducing extra option like 'none' will not break anthing or affect anyone who use codeigniter in their projects.

Feature request 02
============
Currently $this->db-like() and it's related methods using 'WHERE' clause to generate the query. for example:

$this->db->like('title', 'match', 'before');
Will produce: WHERE `title` LIKE '%match' ESCAPE '!'

I'm asking for 'HAVING' clause option for $this->db-like() and it's related methods as a another optional parameter (4th parameter)

for example:
$this->db->like('title', 'match', 'before', 'having');
Should produce: HAVING `title` LIKE '%match' ESCAPE '!'

Reason to request this feature
----------------------------------
'WHERE' clause is not support for aggregate functions like COUNT().

If I need performe search against a aliased field, we have to concatinate text strings here and their to get the job done. may be doing that can introduce security issues.

This is how I do it now:
$this->db->having('title LIKE ', '%match%'); 

But I wish I had a clean way of achieving this without concatinate wildard character around.

**Comment by narfbg** [comment]
01:

> MySQL return rows that seemingly do not match the WHERE clause. Please refer this url to get a good idea about what I'm taling here:
> https://stackoverflow.com/questions/15715719/why-does-mysql-return-rows-that-seemingly-do-not-match-the-where-clause

That's simply bad usage and/or not knowing how to compare by type. PHP does the same thing too - if you compre a string against an integer, the former will be cast to the latter.

... but nevertheless, the 'none' option already exists and does exactly what you suggest.

Before opening issues, please do some basic research.

---

On 02:

> 'WHERE' clause is not support for aggregate functions like COUNT().

This is not true.

> If I need performe search against a aliased field, we have to concatinate text strings here and their to get the job done. may be doing that can introduce security issues.

You don't need to search against aliases, nor is string interpolation necessary for the kind of comparisons you're talking about. There may be a thing to be said about the *convenience* of comparing against aliased names, but for some reason that's not the argument you're making ... Just be honest, don't try to make up sophisticated claims that you can't back up.

The purpose of HAVING is to filter through already selected data instead of scanning entire tables and/or indexes for a subset that you've already fetched. Aliases play a role in that, but aren't the reason why HAVING exists.  
In other words, your argument is again based on misusage.

That being said, HAVING variances of `like()` would have valid appliances and if you come up with a good PR introducing them, I'll be happy to review it.

**Comment by ko2in** [comment]
> ... but nevertheless, the 'none' option already exists and does exactly what you suggest.

The source code does, but the [documentation](https://www.codeigniter.com/userguide3/database/query_builder.html#looking-for-similar-data) doesn't mention about 'none' option.

It mentions only 'before', 'after' and 'both' which is default. I didn't know about that either until I've read your comment. Most of CI users are unlikely to know that unless they read the source code.

Can you please update the documentation?

**Comment by narfbg** [comment]
#5583
